File: tools/foreplay_manager.py
"""
tools/foreplay_manager.py — Project Antigravity
Connection manager that auto-switches between Foreplay API and scraper fallback.
Silent fallback — no user-facing error unless BOTH methods fail.
"""
import os
import sys
import time
from typing import Optional

sys.path.insert(0, os.path.dirname(__file__))
from foreplay_api import test_connection as test_api, get_ad_details
from foreplay_scraper import get_session_cookies_sync, scrape_share_url_sync

# State
_use_scraper = False
_scraper_cookies: dict = {}
_cookies_timestamp: float = 0
COOKIE_REFRESH_INTERVAL = 3 * 3600  # 3 hours (cookies expire at ~4hrs)


def init_connection() -> bool:
    """
    Initialize Foreplay connection.
    Tests API first, falls back to scraper if API is unavailable.

    Returns:
        True if at least one method is available
    """
    global _use_scraper, _scraper_cookies, _cookies_timestamp

    # Try API first
    api_ok = test_api()
    if api_ok:
        _use_scraper = False
        print("✅ Using Foreplay API (primary)")
        return True

    # Fall back to scraper
    print("⚠️ Foreplay API unavailable → initialising scraper session")
    _use_scraper = True
    _scraper_cookies = get_session_cookies_sync()
    _cookies_timestamp = time.time()

    if _scraper_cookies:
        print("✅ Using Foreplay Scraper (fallback)")
        return True

    print("❌ BOTH Foreplay API and Scraper failed")
    return False


def _refresh_cookies_if_needed():
    """Refresh scraper cookies if they're about to expire."""
    global _scraper_cookies, _cookies_timestamp
    if time.time() - _cookies_timestamp > COOKIE_REFRESH_INTERVAL:
        print("🔄 Refreshing Foreplay scraper cookies...")
        _scraper_cookies = get_session_cookies_sync()
        _cookies_timestamp = time.time()


def get_foreplay_ad(
    ad_id: Optional[str] = None,
    share_url: Optional[str] = None,
) -> dict:
    """
    Get ad details from Foreplay (auto-selects API or scraper).

    Args:
        ad_id: Foreplay ad ID (for API)
        share_url: Foreplay share URL (for scraper)

    Returns:
        Dict with videoUrl, landingPageUrl, and other metadata
    """
    global _use_scraper

    if not _use_scraper and ad_id:
        # Try API
        result = get_ad_details(ad_id)
        if result:
            return result
        # API failed for this request — try scraper
        if share_url:
            print(f"⚠️ API failed for {ad_id}, trying scraper...")

    if share_url:
        _refresh_cookies_if_needed()
        return scrape_share_url_sync(share_url, _scraper_cookies)

    return {"videoUrl": "", "landingPageUrl": ""}


def get_connection_status() -> dict:
    """Get current connection status."""
    return {
        "method": "scraper" if _use_scraper else "api",
        "cookiesAge": int(time.time() - _cookies_timestamp) if _use_scraper else None,
        "hasCookies": bool(_scraper_cookies) if _use_scraper else None,
    }


if __name__ == "__main__":
    init_connection()
    print(f"Status: {get_connection_status()}")

File: src/lib/apis/foreplay-manager.ts
/**
 * lib/apis/foreplay-manager.ts — Project Antigravity
 * Foreplay connection manager: API primary + scraper fallback
 */
import { execSync } from 'child_process';
import path from 'path';

let useScraper = false;
let scraperCookies: Record<string, string> | null = null;

const FP_BASE = 'https://api.foreplay.co/v1';

function getFpHeaders(): Record<string, string> {
  return {
    Authorization: `Bearer ${process.env.FOREPLAY_API_KEY || ''}`,
    'Content-Type': 'application/json',
  };
}

export async function testForeplayConnection(): Promise<boolean> {
  const key = process.env.FOREPLAY_API_KEY;
  if (!key) return false;

  try {
    const res = await fetch(`${FP_BASE}/me`, {
      headers: getFpHeaders(),
      signal: AbortSignal.timeout(10000),
    });
    if (res.status === 401 || res.status === 403) {
      console.warn('Foreplay API key invalid or quota exceeded — switching to scraper fallback');
      return false;
    }
    return res.ok;
  } catch {
    return false;
  }
}

async function getForeplayScraperCookies(): Promise<Record<string, string>> {
  try {
    const scriptPath = path.join(process.cwd(), 'tools', 'foreplay_scraper.py');
    const result = execSync(`python -c "
import sys; sys.path.insert(0, '${path.dirname(scriptPath).replace(/\\/g, '\\\\')}')
from foreplay_scraper import get_session_cookies_sync
import json
cookies = get_session_cookies_sync()
print(json.dumps(cookies))
"`, { timeout: 30000, encoding: 'utf-8' });
    return JSON.parse(result.trim());
  } catch (e) {
    console.error('Failed to get scraper cookies:', e);
    return {};
  }
}

export async function initForeplayConnection(): Promise<void> {
  const apiWorking = await testForeplayConnection();
  if (!apiWorking) {
    console.log('Foreplay API unavailable → initialising scraper session');
    useScraper = true;
    scraperCookies = await getForeplayScraperCookies();
  }
}

export async function getForeplayAdDetails(shareUrl: string): Promise<{
  videoUrl: string;
  landingPageUrl: string;
}> {
  if (useScraper && scraperCookies) {
    // Use Python scraper via subprocess
    try {
      const scriptDir = path.join(process.cwd(), 'tools').replace(/\\/g, '\\\\');
      const result = execSync(`python -c "
import sys, json; sys.path.insert(0, '${scriptDir}')
from foreplay_scraper import scrape_share_url_sync
cookies = ${JSON.stringify(scraperCookies)}
result = scrape_share_url_sync('${shareUrl}', cookies)
print(json.dumps(result))
"`, { timeout: 45000, encoding: 'utf-8' });
      return JSON.parse(result.trim());
    } catch (e) {
      console.error('Scraper failed:', e);
      return { videoUrl: '', landingPageUrl: '' };
    }
  }

  // API path
  try {
    // Extract ad ID from share URL
    const match = shareUrl.match(/\/ads\/([a-zA-Z0-9]+)/);
    if (!match) return { videoUrl: '', landingPageUrl: '' };

    const res = await fetch(`${FP_BASE}/ads/${match[1]}`, {
      headers: getFpHeaders(),
      signal: AbortSignal.timeout(15000),
    });

    if (res.ok) {
      const data = await res.json();
      return {
        videoUrl: data.videoUrl || '',
        landingPageUrl: data.landingPageUrl || '',
      };
    }
    return { videoUrl: '', landingPageUrl: '' };
  } catch {
    return { videoUrl: '', landingPageUrl: '' };
  }
}

export function getConnectionStatus(): { method: string; hasCookies: boolean } {
  return {
    method: useScraper ? 'scraper' : 'api',
    hasCookies: useScraper ? !!scraperCookies && Object.keys(scraperCookies).length > 0 : false,
  };
}
